[WIP] Filestore Implementation

kevina opened the pull request:
Closes Issue #875, avoid duplicating files added to ipfs

**NOT READY FOR MERGE**

Rebased #2600 on master.

Quicklinks: [Code](https://github.com/ipfs-filestore/go-ipfs), [README](https://github.com/ipfs-filestore/go-ipfs/blob/master/filestore/README.md), 

TODO to get this merged:
- [ ] Rebase on master turning the chain of commits into a reasonable ChangeSet
- [ ] Agree on major infrastructure changes
  - [ ] Multi-blockstore (#3119)
- [x] Separate out non-filestore bits of infrastructure change into there own pull request
  - Multi-blockstore: #3257
  - DAGService.GetLinks: ~~#3255~~
  - Others: see milestone [Filestore implementation](https://github.com/ipfs/go-ipfs/milestone/26)
- [ ] Code review
- [ ] Merge

Note The filestore is very basic right now, but it is functional.  I will likely continue to improve and submit new pull requests for the enhanced functionally but right now I fell it is important a basic implementation gets in so that it will get used, it can be labeled as an experimental feature and disabled by default, but available for those that want to use it.  I consider the code production ready.


Resolves #875

whyrusleeping commented on blocks/blocks.go:
```diff
@@ -19,48 +20,61 @@ type Block interface {
 
-// Block is a singular block of data in ipfs
-type RawBlock struct {
+type BasicBlock struct {
```
aww, renaming this from `RawBlock` to `BasicBlock` makes the diff so noisy... I was hoping to avoid that by getting that change merged separately

kevina replied:
Sorry, I missed that change as it was in a separate commit.  Let me know if you know if you want me to do anything about it, like submitting this change separately.

kevina replied:
@whyrusleeping, this rename is only in one file, it is exported because I might eventually refactor my Filestore specific changes to blocks/blocks.go into another file.

kevina commented:
I image that this is going to take a while before it gets merged.  So for now I think it will be best to maintain this as a separate fork while the goal of eventually merging everything in.

@whyrusleeping I will still like to work with you on API issues and work to get major API changes merged first to avoid stagnation.

I have created a README for the new filestore available here https://github.com/ipfs-filestore/go-ipfs/blob/kevina/filestore/filestore/README.md some notes on my fork are available here https://github.com/ipfs-filestore/go-ipfs/wiki.

@jefft0 if you are interested I could use some serious testing now.  Others all of course welcome to test it out.

whyrusleeping commented on commands/files/readerfile.go:
```diff
@@ -13,10 +13,12 @@ type ReaderFile struct {
 
-func NewReaderFile(filename, path string, reader io.ReadCloser, stat os.FileInfo) *ReaderFile {
-	return &ReaderFile{filename, path, reader, stat}
+func NewReaderFile(filename, path, abspath string, reader io.ReadCloser, stat os.FileInfo) *ReaderFile {
```
this can probably be changed so that it only accepts the absolute path instead of a filename, path and absolute path.

whyrusleeping replied:
(and a similar change can be made throughout this part of the codebase)

kevina replied:
@whyrusleeping  I added absolute path, because I was unsure how FullPath was used in the code base.  Is it acceptable for full path to be an absolute path?  I am unsure how FullPath and "add -r" interact.

whyrusleeping replied:
To be honest, i'm not entirely sure either. But that said, our tests in that area are quite thorough, If you change something, go into `test/sharness`, run `make deps` and then run `./t0040-add.sh`, and maybe some of the other `t004*` tests too to make sure nothing weird happened

kevina replied:
All right I will start by trying to change FullPath to an absolute path add see what happens when used with "add -r".  I am will make sure the contents of the directory objects are correct.

kevina replied:
It turns our that when adding directories the directory is part of the filename.  For example after adding some debug output here is the result of one of the test:

```
expecting success: 
        mkdir mountdir/planets &&
        echo "Hello Mars!" >mountdir/planets/mars.txt &&
        echo "Hello Venus!" >mountdir/planets/venus.txt &&
        ipfs add -r mountdir/planets >actual

Filename: planets/mars.txt
Path: mountdir/planets/mars.txt
Filename: planets/venus.txt
Path: mountdir/planets/venus.txt
ok 42 - 'ipfs add -r' succeeds
```

Thus we still need to pass in both a filename and a path. It seams okay to just make the path absolute, so I will do that I remove the change that adds an absolute path to the various parts in the `files` package.

whyrusleeping replied:
We could extract the filename from the passed in path though

kevina replied:
I tried that and it is not going to work.  This test will fail.  The filename needs to be "planets/mars.txt" if you try to extract it it will be just "mars.txt".   You will get this:

```
expecting success: 
        mkdir mountdir/planets &&
        echo "Hello Mars!" >mountdir/planets/mars.txt &&
        echo "Hello Venus!" >mountdir/planets/venus.txt &&
        ipfs add -r mountdir/planets >actual

ok 42 - 'ipfs add -r' succeeds

expecting success: 
        PLANETS="QmWSgS32xQEcXMeqd3YPJLrNBLSdsfYCep2U7CFkyrjXwY" &&
        MARS="QmPrrHqJzto9m7SyiRzarwkqPcCSsKR2EB1AyqJfe8L8tN" &&
        VENUS="QmU5kp3BH3B8tnWUU2Pikdb2maksBNkb92FHRr56hyghh4" &&
        echo "added $MARS planets/mars.txt" >expected &&
        echo "added $VENUS planets/venus.txt" >>expected &&
        echo "added $PLANETS planets" >>expected &&
        test_cmp expected actual

> diff -u expected actual
--- expected    2016-05-11 03:20:12.742743658 +0000
+++ actual      2016-05-11 03:20:12.466738156 +0000
@@ -1,3 +1,2 @@
-added QmPrrHqJzto9m7SyiRzarwkqPcCSsKR2EB1AyqJfe8L8tN planets/mars.txt
-added QmU5kp3BH3B8tnWUU2Pikdb2maksBNkb92FHRr56hyghh4 planets/venus.txt
-added QmWSgS32xQEcXMeqd3YPJLrNBLSdsfYCep2U7CFkyrjXwY planets
+added QmPrrHqJzto9m7SyiRzarwkqPcCSsKR2EB1AyqJfe8L8tN mars.txt
+added QmU5kp3BH3B8tnWUU2Pikdb2maksBNkb92FHRr56hyghh4 venus.txt

not ok 43 - 'ipfs add -r' output looks good
#
#               PLANETS="QmWSgS32xQEcXMeqd3YPJLrNBLSdsfYCep2U7CFkyrjXwY" &&
#               MARS="QmPrrHqJzto9m7SyiRzarwkqPcCSsKR2EB1AyqJfe8L8tN" &&
#               VENUS="QmU5kp3BH3B8tnWUU2Pikdb2maksBNkb92FHRr56hyghh4" &&
#               echo "added $MARS planets/mars.txt" >expected &&
#               echo "added $VENUS planets/venus.txt" >>expected &&
#               echo "added $PLANETS planets" >>expected &&
#               test_cmp expected actual
#
```

Here is a quick patch to try, it might not apply cleanly for you though.

```
diff --git a/commands/files/readerfile.go b/commands/files/readerfile.go
index 2508fe1..d7f5443 100644
--- a/commands/files/readerfile.go
+++ b/commands/files/readerfile.go
@@ -4,6 +4,7 @@ import (
        "errors"
        "io"
        "os"
+       gopath "path"
 )
 
 // ReaderFile is a implementation of File created from an `io.Reader`.
@@ -17,7 +18,11 @@ type ReaderFile struct {
        baseInfo ExtraInfo
 }
 
-func NewReaderFile(filename, path string, reader io.ReadCloser, stat os.FileInfo) *ReaderFile {
+func NewReaderFile(_, path string, reader io.ReadCloser, stat os.FileInfo) *ReaderFile {
+       filename := path 
+       if path != "" {
+               filename = gopath.Base(path)
+       }
        return &ReaderFile{filename, path, reader, stat, 0, PosInfo{0, path}}
 }
```

whyrusleeping commented on repo/fsrepo/defaultds.go:
```diff
@@ -13,22 +14,30 @@ import (
+var _ = io.EOF
 
-func openDefaultDatastore(r *FSRepo) (repo.Datastore, error) {
+func openDefaultDatastore(r *FSRepo) (repo.Datastore, *filestore.Datastore, error) {
```
I'm not sure on this, but i think that this might be cleaner if we construct the filestore outside of this function, and then pass it in as a parameter. Doing so would move the `useFileStore` 'global' check up a call or two, making it easier to abstract that out.

kevina replied:
The useFileStore is a temporary hack.  I thought about making it a config var., but once a filestore is enabled in the repository, disabling it can cause problems, especially with the pinner.   I am still not sure but am not thinking that it should be enabled via a command.

whyrusleeping commented on Godeps/_workspace/src/github.com/ipfs/go-datastore/multi/multi.go:
```diff
@@ -0,0 +1,204 @@
+
+// Note: The advance datastore is at index 0 so that it is searched first in Get and Has
+
+func New(adv ds.Datastore, normal ds.Datastore, aux []ds.Datastore, roAux []ds.Datastore) *Datastore {
```
We should make a 'type switched datastore' with this. That way we can say things like "store all `[]byte` in datastore X" and "store all `*filestore.FileStore` in datastore Y"

I did a sketch to check if this was possible (it is, yay!)
```go
package main

import (
	"fmt"
	"reflect"
)

type X struct{}
type Y struct{}

func main() {
	v := make(map[reflect.Kind]string)

	set := func(i interface{}, val string) {
		v[reflect.TypeOf(i).Kind()] = val
	}

	get := func(i interface{}) string {
		return v[reflect.TypeOf(i).Kind()]
	}

	set("test", "A")
	set(&X{}, "B")
	set(&Y{}, "C")

	fmt.Println(get(&Y{}))

}
```

whyrusleeping commented:
@kevina Yeah, theres a LOT here, so its going to take some time. I don't disagree with you maintaining a fork in the meantime. That said, as I find time my plan is to pick out parts of the code that can be merged up easily. I think the next candidate for that is the commands/files code refactor to use absolute paths.

kevina commented:
@whyrusleeping,  One idea I had which might avoid some special casing is to replicate what is done with `add --no-copy`
```
                if hash {
                        nilnode, err := core.NewNode(n.Context(), &core.BuildCfg{
                                //TODO: need this to be true or all files
                                // hashed will be stored in memory!
                                NilRepo: true,
                        })
                        ...
                        n = nilnode
                }
```
This could help avoid special case code in at least a few places.  It will at minimal avoid the special case code in the blockstore package as I can use my own implementation.

Do you think this is an idea worth pursuing?  I image the special node will share many of the data structures from the real node to avoid any data race or related issues.

whyrusleeping commented:
@kevina I think that might be a good idea, my only concern is that in order to read the blocks out of the filestore would you still need that custom created node?

kevina commented:
@whyrusleeping, no the custom node will be for adding only.  We will still need some sort of multi-datastore for reading blocks.  See my comments for #2747.

kevina commented:
@whyrusleeping there is a lot going on in an  Ipfs Node.  What I need is not a mock node like the code that `--only-hash` uses but rather create a new Node that acts like a view into the original node except that add's are handled differently.  I was not confident I could pull this if with a full Ipfs Node.

With that in mind, I refactored the adder code a little to bundle all the needed services into the `DataServices` struct see commit c41a47e.  A proper interface that exports just methods needed by the adder would be nice, but I think this is a good start.  This new interface could be used to simply a lot of code elsewhere, including the code for just computing the hashes, but I don't fully understand what is going on there so I left the code for `add --hash-only` alone.

I use this new interface to to create a DataServices view in commit e6bfbfa that has uses an alternative Blockstore and leaves the existing Blockstore and caching Blockstore wrapper alone.  In commit b54f2f4 I factor out the filestore specific code from the DagServices and instead install small hooks. 

In commit f9a51b1 I completely eliminated the AdvReader code you didn't like and instead computed the offset's in the DAG builder code (the Balanced builder to be specific, the Trickle builder doesn't support the filestore just yet).

All in all I cleaned up a lot of the special case code.

I you are okay with the changes in c41a47e I can separate that out into it's own pull request.

whyrusleeping commented:
@kevina from irc, we can move towards making the `NewAdder` signature be: `NewAddr(context.Context, bs bstore.Blockstore, pin pinning.Pinner, ds dag.DAGService)` which eliminates the need for the partial initialization stuff

kevina commented:
@whyrusleeping f5fb3c9 basically does what you ask for.

kevina commented:
For anyone who has been using my code, I just pushed some major changes to the filestore.  Have a look at the README.  There is also a sample script to to add the contents of a directory to the filestore and keep the directory in sync with the filestore.

At this point I could really use some testers.

jefft0 commented:
HI Kevin. I've just finished preparing my big store of webcam files. I should be able to pull your new code and try it tomorrow.

kevina commented:
@jefft0 thanks.  Le me know how it goes.  You should be able to add files while the server is online so that should make using it easier.

jefft0 commented:
@kevina, I'm on Windows 8.1 because my webcam software (and attached storage) is for Windows. I was able to "make install" in Cygwin. I tried the following in Cygwin:

    ipfs filestore add -P -r /cygdrive/f/cameras/2016

It says "The system cannot find the path specified." I also tried the following in a Windows command prompt:

    ipfs filestore add -P -r F:\cameras\2016

It says "File path must be absolute." Thinking that maybe it doesn't like Windows drive letters. I changed to the F drive and just tried:

    ipfs filestore add -P -r \cameras\2016

with the same result "File path must be absolute."

Sorry to stick you with Windows problems, but this is the system where I can do serious stress testing. Any thoughts?

kevina commented:
@jefft0 try dropping the -P, that is just use `ipfs filestore add -r /cygdrive/f/cameras/2016` or try `ipfs filestore add -r F:\cameras\2016` if that does not work, let's take this to an issue on my fork: https://github.com/ipfs-filestore/go-ipfs/issues

jefft0 commented:
I tried that. Same problem. I opened an issue https://github.com/ipfs-filestore/go-ipfs/issues/6 . (I copied the webcam files to the C drive to remove doubts about using the external F drive.)

jefft0 commented:
A UI issue. I did `ipfs filestore add -r c:\public\cameras` on a video collection of about 12 TB. The progress bar only counted 1.32 TB. Indeed, after about a day, the progress bar shows progress over 100%:

     1.45 TB / 1.34 TB [==============================] 108.61% -59m6s

So I think it is processing all the files, but something is wrong with how it makes the initial total. I tried it on other large collections of different sizes over 2 TB and each time the total shows 1.34 TB.

jefft0 commented:
Let's move discussion of the progress bar total to https://github.com/ipfs-filestore/go-ipfs/issues/8 .

kevina commented:
@whyrusleeping, I think I figured out how to get ride of the `DataPtr` that you didn't like, but I am not 100% sure I like the changes.  I pushed the changes on the [ipfs-filestore:no-altdata](https://github.com/ipfs/go-ipfs/compare/master...ipfs-filestore:no-altdata) branch and have not pushed the changes (see https://github.com/ipfs-filestore/go-ipfs/compare/d3bf502...ipfs-filestore:no-altdata) to this pull request yet.

The filestore requires an alternative encoding of the protocol buffer that doesn't contain the Unixfs Data field (i.e. the data from the file).  Previously the `DataPtr` was used to build this alternative encoding.  In the latest change I basically don't worry about it and then just reencode the protocol buffer and remove the Data field just before sending it to the filestore.  Now the only thing that is stored in the `merkledag.Node` is the `PosInfo`, all the code related to the `DataPtr` and `AltData` is gone.

I originally avoided this because I thought it would be slow.  It might still be a little, but not noticeably so in my very informal tests.

Let me know what you think.  Also If you are okay with storing `PosInfo` in `blocks.Block` I can eliminate the hooks in  [merkledag/merkledag.go](https://github.com/ipfs/go-ipfs/compare/master...ipfs-filestore:no-altdata#diff-f8ac90e3d5339cd2475cd5b3ec8614fd).  I will instead move all the logic to my alternative blockstore.  

Thanks

whyrusleeping commented:
@kevina That looks much cleaner, I think that moving the `PosInfo` into the blocks is a good idea and I doubt that we will see a perf hit from doing things this way. (especially since we're comparing perf relative to disk writes and syncs).

Would you like to schedule a group hangout sometime to talk more about this PR? We've had a lot of increased interest lately in this feature from a few different groups. I think it makes sense for us to catch up and get a good plan forward on getting this in :)

cc @jbenet @lgierth @Kubuxu

kevina commented:
@whyrusleeping, okay I pushed the changes from the no-altdata onto this pull request.  I also moved PosInfo into blockstore and eliminated the hooks in `merkledag/merkledag.go`.

As we discussed over IRC I will write up some sort of document on the major issues that need to be resolved before I am comfortable merging this in and we can go from there.

kevina commented:
@whyrusleeping I posted a draft document here https://github.com/ipfs-filestore/go-ipfs/wiki/Landing, please look it over and let me know what to do with it.  In particular I stated to discuss the first issue "Need Some Sort of Support For Multiple Datastores" in #2747, should we keep the discussion there (I can update the description) or is there a better place.  Thanks.

kevina commented:
@whyrusleeping is there anything I can do at this point to move this forward?

whyrusleeping commented:
@kevina I need to get some more review in on this. In the meantime, if you want to start hacking on the 'best effort' pins, that would be helpful

RichardLitt commented on core/commands/filestore.go:
```diff
@@ -0,0 +1,946 @@
+	Helptext: cmds.HelpText{
+		Tagline: "Add files to the filestore.",
+		ShortDescription: `
+Add contents of <path> to the filestore.  Most of the options are the
```
Use single quotes instead of ".

kevina replied:
Where do you want me to use single quotes?

RichardLitt replied:
for 'ipfs add'

kevina replied:
Okay, that is what I thought but wasn't sure.  Will fix.

RichardLitt commented on core/commands/filestore.go:
```diff
@@ -0,0 +1,946 @@
+		logical, _, _ := req.Option("logical").Bool()
+		physical, _, _ := req.Option("physical").Bool()
+		if logical && physical {
+			return errors.New("Both --logical and --physical can not be specified.")
```
Errors shouldn't start with a capital or end with a period (sorry).

kevina replied:
Is this a policy?  Is this documented somewhere.

Kubuxu replied:
It think it is general Go style but we might want to note that somewhere.

RichardLitt commented on core/commands/filestore.go:
```diff
@@ -0,0 +1,946 @@
+	opts = append(opts, AddCmd.Options...)
+	opts = append(opts,
+		cmds.BoolOption("server-side", "S", "Read file on server."),
+		cmds.BoolOption("l", "logical", "Create absolute path using the PWD from environment."),
```
Write out PWD? I don't know what this stands for.

kevina replied:
`PWD` is a environmental variable from the shell, spelling out would be incorrect.

RichardLitt commented on core/commands/filestore.go:
```diff
@@ -0,0 +1,946 @@
+var lsFileStore = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "List objects in filestore",
+		ShortDescription: `
```
Add synopsis.

Kubuxu replied:
We have synopsis auto-generator in the pipeline (https://github.com/ipfs/go-ipfs/pull/2785) so I don't think it is needed.

RichardLitt commented on core/commands/filestore.go:
```diff
@@ -0,0 +1,946 @@
+
+var lsFileStore = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "List objects in filestore",
```
Needs period.

jefft0 commented:
Stress test results: On my Windows 8.1 desktop,I did `ipfs filestore add -r` on a video collection of 12.5 TB, total processing hours of 115 hours. The sizes of my `.ipfs` folders are:

    blocks       175 MB
    datastore    916 KB
    filestore-db 9.66 GB (5,333 ldb files)

RichardLitt commented on core/commands/filestore.go:
```diff
@@ -0,0 +1,946 @@
+If --quiet is specified only the hashes are printed, otherwise the
+fields are as follows:
+  <hash> <type> <filepath> <offset> <size> [<modtime>]
+where <type> is one of"
```
Should be `one of:`?

kevina replied:
Yes will fix.

RichardLitt commented on core/commands/filestore.go:
```diff
@@ -0,0 +1,946 @@
+
+		if !more {
+			if w.checksFailed {
+				w.errs = append(w.errs, "Some checks failed.")
```
Change to: `some checks failed` (no period, decapitalize)

RichardLitt commented on core/commands/filestore.go:
```diff
@@ -0,0 +1,946 @@
+The normal output is:
+  <status> <hash> [<type> <filepath> <offset> <size> [<modtime>]]
+where <hash> <type>, <filepath>, <offset>, <size> and <modtime>
+are the same as in the "ls" command and <status> is one of
```
are the same as in the 'ls' command and <status> is one of:

RichardLitt commented on filestore/README.md:
```diff
@@ -0,0 +1,207 @@
+it from another node such as the ipfs.io gateway.
+
+Paths stores in the filestore must be absolute.  You can either
+provide an absolute path or use one of `-P` (`--physical`) or -l
```
`-l`

RichardLitt commented on filestore/README.md:
```diff
@@ -0,0 +1,207 @@
+Paths stores in the filestore must be absolute.  You can either
+provide an absolute path or use one of `-P` (`--physical`) or -l
+(`--logical`) to create one.  The `-P` (or `--physical`) means to make
+a absolute path from the physical working directory without any
```
an absolute path

RichardLitt commented on filestore/README.md:
```diff
@@ -0,0 +1,207 @@
+provide an absolute path or use one of `-P` (`--physical`) or -l
+(`--logical`) to create one.  The `-P` (or `--physical`) means to make
+a absolute path from the physical working directory without any
+symbolic links in it; the -l (or `--logical`) means to use the `PWD`
```
Put ` around -l

RichardLitt commented on filestore/README.md:
```diff
@@ -0,0 +1,207 @@
+all files in a directly to the filestore and keep the filestore in
+sync with what is the directory.  Just specify the directory you want
+to add or update.  The first time it is run it will add all the files
+in the directory.  When run again it will readd any modified files.  A
```
readd

kevina replied:
There is evidence for "readd" being a word: http://www.dictionary.com/browse/readd, https://en.wiktionary.org/wiki/readd.  I would rather leave it in unless there is a strong objection.  I can change "readd" to "re-add" if you think that is more correct through.

RichardLitt replied:
Ha! I thought this was a mispelling of 'read'. Keep it in. Maybe use a hypen: re-add?

kevina replied:
Okay.  I will use a hyphen to avoid the confusion.

RichardLitt commented:
@kevina Thanks for being patient - just trying to make this conform to the rest of the codebase. :)

kevina commented:
@whyrusleeping @Kubuxu since it looks like we are moving towards using base32 keys I am thinking about killing the datastore level in my filestore and just directly implementing a blockstore as I am not sure they is any benefit to implementing it as a datastore for this.  The "datastore" I implement is special cased for IPFS nodes anyway and will serve little use as a general purpose datastore, so what's the point.  Thoughts?

kevina commented:
The merging of #2903 will create a non-backwards completable change to the filestore once I merge with master.  I am undecided on how to handle the upgrade.

@jefft0 (and anyone else) are you using the filestore in a (semi) production environment or just for testing?  The easiest path would be to require that all files are simply readded, although with a little effort I could do something better that won't require rereading the file contents in most cases.  Thanks.

jefft0 commented:
@kevina, don't let me slow you down. I realize we're at a testing stage. I can take another 115 hours and re-add the files.

kevina commented:
@whyrusleeping it turns out that @Kubuxu bloom filter significantly complicates things.  In order to even merge master I need to implement some form of multi-datastore support (or just disable the Bloom filter in my branch as a temporary hack).  The current hacks I have in there now won't cut it as the bloom filter will not recognize the content in the filestore.

I realize that everyone is busy, but I would like to see some forward progress in working through the issues outlined at: https://github.cohttps://github.com/ipfs-filestore/go-ipfs/wiki/Landing.  If it will help move things forward we could do a group hangout with interested parties.  Note that for me it will need to be voice only.

I am very tempted just to implement my plan for Multi-datastore support just so that I can merge with master, but I am hesitant to do so as I already made a number of design choices without really getting support from other core IPFS community members.

jefft0 commented:
> Are you using the filestore in a (semi) production environment or just for testing? The easiest path would be to require that all files are simply readded, although with a little effort I could do something better that won't require rereading the file contents in most cases.

@kevina, When do you plan to merge the change to the filestore format?

kevina commented:
@jefft0 right now I am waiting for feedback from @whyrusleeping on my last comment (concerning the bloom filter).  If enough times passes without any feedback I could hack something by disabling the bloom filter but I rather not.

whyrusleeping commented:
@kevina sorry for taking so long to get back to you. It has been a very hectic three weeks and i'm just now getting back in the swing of things on github. 

I think you are probably good to start implementing the 'multi datastore' idea. Take a look at the [tiered datastore](https://github.com/ipfs/go-datastore/blob/master/tiered/tiered.go) we have already. Its not quite what you want, but it might be a good model to start with.

The abstractions should be set up such that any layering of blockstores are able to correctly handle the blocks being stored.

Responding to some of your points in your filestore proposal doc:

- For the garbage collector, I think that adding a `GetLinks` method to the dagservice isnt a bad idea. It would as you say be pretty useful in many situations (the resolver could use this optimized method to speed up graph traversals)

- on invalid blocks, We probably do want to try and continue on errors. This is a difficult problem to solve, we need to change the API so that those relevant calls can return errors alongside their other results, and we also need to figure out what a proper return code for such a command result would be. Does a command that enounters invalid blocks exit 'successfully'?

- Re implementing the filestore as a blockstore, This is an interesting idea. I'm not sure how everything would fit together with that. If you think thats a good way forward i'd like to see a short writeup on how that would be constructed.

kevina commented:
@jefft0 I merged the changes in the filestore format and created a way to upgrade the filestore.  The upgrade code turned out to be really simple.

First upgrade the repo (it can be trigged automatically if you use `ipfs daemon`).  After the repo is upgraded run `ipfs filestore upgrade` to fix the keys in the filestore that are in the original format. It is best to do this with the daemon shutdown (it could in theory work with the daemon running, but this has not been tested).

If a key is mangled it will need to read the contents from the file to rehash.  If it is unable to do so for some reason it will output a warning and use the mangled hash.  You can inspect the managed hash with `ipfs filestore ls` and it can be manually removed using `ipfs filestore rm`.  (`ipfs clean ...` will also likely remove it).

If the process is interrupted, don't worry, just run it again and it will pick up from where it left off.

If you have any problems please file a issue at https://github.com/ipfs-filestore/go-ipfs/issues.

kevina commented:
Also note that I have not yet fixed the filestore to work with the Bloom filters.  This should not be a problem though as the Bloom filters are now disabled by default.

jefft0 commented:
`ipfs filestore upgrade` has been running for a day now. Is there any way to find out the progress?

kevina commented:
@jefft0 I am really surprised it is taking that long, sorry I didn't think about including some sort of progress indicator.  You could try stopping it than do a `ipfs filestore ls`.  If an entry is not converted the key will start with '???`.  You can then restart it and it should pick up from where it left off.

jefft0 commented:
@kevina Checking for ??? is good enough. I stopped `ipfs filestore upgrade` and `ipfs filestore ls` showed only 2% remaining with ???. I restarted `upgrade` and it kept running, so I stopped it again and `ls` shows all converted with no ???. But when I start `upgrade` again, it keeps running anyway. I will stop it and use the filestore as normal to see if I come across any problems.

kevina commented:
Sorry, `ipfs upgrade` is very simple.  It will always run, it just won't do anything if there is nothing to upgrade.  As long as you have no '???' you should be okay.

jefft0 commented:
@kevina, should I be testing with these latest commits, or wait for more commits?

kevina commented:
@jefft0 I would wait, some more commits are pending

kevina commented:
@whyrusleeping Thanks for your feedback.  The filestore is now in a state where it has sensible semantics with regards to the rest of IPFS.

I decided to go with a multi-blockstore but to keep the filestore as a datastore.

Many of the details or in the commit logs, but I will spell them out in greater detail this week.  Once the core developers are in agreement on the semantics I will start to rebase and turn this into a nice set of commits and factor out what I can into separate pull requests.

@jefft0 (and others) now would be a good time for some more testing.  There are still some rough edges but things should be usable.  The core functionally is the same, but a lot of the details have changed so might want have another look at the README.  (There are no changes to repo.)

One think of note is I no longer pin objects in the filestore.   If you added the files using "-r" the directory object got pinned and hence everything under it is indirectly pinned.  This will cause problems if you ever need to remove files so if you don't care about the directory object you might want to unpin it.

jefft0 commented:
> if you don't care about the directory object you might want to unpin it.

What are the best commands to find pinned directories and to unpin them?

kevina commented:
@jefft0 Something like:
```
ipfs pin ls --type=recursive -q # to find pins
ipfs ls <HASH>                  # to discover the contents
ipfs pin rm <HASH>              # to remove the pins
```
Hopefully you don't have too many pins.  If not you do you could do some basic scripting using the Unix `sh` to dump the contents of all the pins in a readable fashion.
```
for hash in `ipfs pin ls -q --type=recursive`; do echo "*** $hash"; ipfs ls $hash; done
```
If any pins are a directory, it will show something like
```
*** QmWczekSRtsGdqJGASEgb2WPMApFUY9Epa7HLhzfY3vNnP
QmQUZzkwP6aLEyuhZbzkNznfwhV59e4RmjN7PbH89Fc5pS 3316736  12Dicts/
QmSbBy1izdeC43KRBPK5jEiVRshS9GboLWhL44F2YVMYgY 15998763 AGID/
QmccKNwQc6VVxkCPVtM5hEZWbYt2dRdyspKAw4T7xLg7tP 8072754  Alt12Dicts/
```

jefft0 commented:
Thanks for the help. I unpinned my files and directories. It makes sense that filestore blocks are not pinned. Is there a way to pin the directory that contains them (maybe recursively) without pinning the filestore files it contains?

kevina commented:
@jeff0 you can use a direct pin for each directory, but it is not recursive so it will involve a lot of pins.  Anything under the mfs, see "ipfs files", will be now be pinned on a best effort bases so you could copy the directory there using "ipfs files cp".  Currently there is no way to directly create a best-effort pin.

kevina commented:
@whyrusleeping @jbenet I have updated the description with a outline as how I see this pull request getting merged.  We can start with #3119.

kevina commented:
All: Now would be a good time for another round of testing.  There is a new script "add-dir" to aid in keeping a directory in sync.

kevina commented:
Merged with master.  This was a major PITA.  I did it in pieces so I can be sure I handled all the conflicts correctly.  There was one breaking change (see #3253) and the switch to CidV0 took some work to fix.  @whyrusleeping it would be really nice if we could get the core functionally in ASAP.

kevina commented:
All: Sometime within the next week I am likely to start redoing the commits.  In process I am likely to remove any backwards compatibility bits in the code.  If you are using this code you should run `filestore upgrade` to make sure it will still work with the new version.  Old versions of the code will still able to be found in the `master` branch of my fork.

kevina commented:
@whyrusleeping (and others), I have separated out the important non-filestore bits into separate pull requests.  Note that many of them will conflict with each other, but I can easily rebase the unmerged ones master to fix the conflicts.

whyrusleeping commented on merkledag/merkledag.go:
```diff
@@ -93,6 +106,24 @@ func (n *dagService) Get(ctx context.Context, c *cid.Cid) (*Node, error) {
+}
+
+func (n *dagService) WithBlockService() bool {
+	return true
```
Shouldnt this return something like `return n.Blockservice != nil` ?

kevina commented:
Note to see what changes remain after all the outstanding pull requests are merged:
```
git fetch https://github.com/ipfs-filestore/go-ipfs.git
git diff --stat 1872685..f4785481 -- . ':!*filestore*' ':!*t026*.sh'
```
This will also exclude any new files created.  The results should be:
```
 blocks/blocks.go                   | 14 +++++++++++++-
 cmd/ipfs/ipfs.go                   |  2 ++
 commands/cli/cmd_suggestion.go     |  4 ++++
 commands/cli/parse.go              | 49 +++++++++++++++++++++++++++----------------------
 commands/files/file.go             | 11 +++++++++++
 core/builder.go                    | 12 +++++++++++-
 core/commands/add.go               | 22 +++++++++++++++++++++-
 core/commands/block.go             | 60 ++++++++++++++++++++++++++++++++----------------------------
 core/commands/root.go              |  1 +
 core/coreunix/add.go               | 26 +++++++++++++++++++++-----
 importer/balanced/builder.go       | 12 +++++++++---
 importer/chunk/rabin.go            | 10 ++++++++--
 importer/chunk/splitting.go        |  5 +++++
 importer/helpers/dagbuilder.go     | 43 ++++++++++++++++++++++++++++++++++++-------
 importer/helpers/helpers.go        | 17 +++++++++++++++--
 merkledag/node.go                  | 11 +++++++++++
 repo/config/config.go              |  1 +
 repo/config/datastore.go           |  5 +++++
 repo/fsrepo/defaultds.go           | 46 ++++++++++++++++++++++++++++++++++++++++++++++
 test/sharness/t0235-cli-request.sh |  2 +-
 20 files changed, 280 insertions(+), 73 deletions(-)
```

atommixz commented:
i need 'ipfs add --no-copy' by default btw.

kevina commented:
@atommixz 
> i need 'ipfs add --no-copy' by default btw.

note, that it is `filestore add` now, and  there is no way to make it "by default".

kevina commented:
I picked out all the non-filestore specific bits I could.  Here is an updated diff:
```
# (git diff --stat next..kevina/filestore  -- . ':!*filestore*' ':!*t026*.sh')
$ git diff --stat 31fcd1855..e7db3e4a1  -- . ':!*filestore*' ':!*t026*.sh'
 cmd/ipfs/ipfs.go                   |  2 ++
 core/builder.go                    |  7 +++++++
 core/commands/add.go               | 23 ++++++++++++++++++++++-
 core/commands/root.go              |  1 +
 repo/config/config.go              |  1 +
 repo/config/datastore.go           | 11 +++++++++++
 repo/fsrepo/defaultds.go           | 46 ++++++++++++++++++++++++++++++++++++++++++++++
 test/sharness/t0235-cli-request.sh |  2 +-
 8 files changed, 91 insertions(+), 2 deletions(-)
```

jefft0 commented:
> Filestore: Support Raw Blocks.

@kevina, should I re-add my files so that they use raw blocks? (Tell me when is a good time because it will take days to re-add.)

kevina commented:
> should I re-add my files so that they use raw blocks? (Tell me when is a good time because it will take days to re-add.

I would wait, the implementation is still experimental.  Raw block support is not even merged into ipfs-go/master yet.  And even when it is there is no rush as the old format will continue to be supported for a good while (at least I think so, @whyrusleeping correct?).

kevina commented:
@jefft0, I should add that one of the advantages of using the filestore is that you can have both formats (the original and with raw leaves) at the same time with little space wastage.

rht commented:
This sure is a feat, pulled singlehandedly on filestore, :clap: :clap:, @kevina! If this PR is carefully splitted into 10 PR's consisting of 10 commits each (if 50% have already been merged), this should be doable within a week (or two the longest).

kevina commented:
Dear All,

For those of you who have not figured out what is going on yet: this p.r. is unlikely to be merged as is and instead the core developers decided on a much more basic implementation.  Please see #3629.

I was not included when the decision was made and am in general not all happy with how the p.r. was handled.  With the way things are going it does not look like most of the functional of this branch will ever make it into IPFS.

I want to thank everyone who has tested and gave me feedback.

There is not much more to say at this point.

Kevin

jefft0 commented:
Hi Kevin. Thanks so much for writing great code and quickly responding to my questions. Your contribution is what made IPFS usable for me. And your determination to remain level-headed in a frustrating situation is admirable.

atommixz commented:
Hi Kevin. Sometimes we write code that will never be used. Shit happens bro. Thanks for a great contribution to the solution of this problem. We waited for the decision of this problem for a very long time. I hope they have done something that we can use. Perhaps Upspin from Google will solve this problem better.

jbenet commented:
Regarding this comment:

> For those of you who have not figured out what is going on yet: this p.r. is unlikely to be merged as is and instead the core developers decided on a much more basic implementation. 
>
> I was not included when the decision was made and am in general not at all happy with how the p.r. was handled. With the way things are going it does not look like most of the functional of this branch will ever make it into IPFS. With the way things are going it does not look like most of the functional of this branch will ever make it into IPFS.

@kevina you're being unfair and you're misleading everyone who wasn't privy to all of our exchanges, many of which happened outside github issues. 

Before you even began this work I warned you (and others) how difficult getting this right would be, and how unlikely it would be for initial solutions to be merged. I discouraged you from expecting that a solution good enough for you would be good enough for everyone, and would make it into master. I discouraged you from working on it under those expectations, and from working on it at all without requisite systems engineering experience. You did it anyway and submitted this and other PRs. 

Many people, myself included, appreciated and applauded your exploration of the design space, and your efforts to reach something pragmatic. Many people, including @whyrusleeping and myself, supported you throughout this process, guided your implementation and spent a lot of our time helping you with it, more time than it would have taken us to implement it.

I apologize for -- on average -- taking a long time to review your PRs and get feedback to you. There were many points in the history of these PRs where I personally was the one who needed to provide feedback, and it took me a very long time to do so. Unfortunately, I am no longer in the position to allocate freely, and my time and bandwidth are spoken for weeks or months in advance. Unfortunately, I had to spend that bandwidth in other ways. I could have done a much better job at providing limited but more frequent feedback, or at least warning you how long it might take me to get back. This is definitely something I need to do better, no matter the mountain of other things i do for the project and the community.

However, I did not think, and still do not, the PRs you presented are a solution that would create a great UX -- it has multiple brittleness problems (eg it can leave thousands or millions of references to other points in the filesystem dangling, it does not play well with mounts that may change location, the UX of encountering those dangling references and fixing them is bad, ...). I repeatedly told you that your solution did not consider important systems expectations, had such brittleness issues, had UX issues, or discounted their importance, and that it did not yet pass my expectations for a robust, usable design we can impose on every single one of our users. I encouraged you to improve on those issues, and encouraged you to maintain your own fork if you chose not to. I also encouraged you to find a way this could be a plug-in functionality, instead of having to maintain a fork with lots of rebasing, nor doing the hard work redesigning and reimplementing to be good enough to get into master. Many times you dismissed my concerns and issues raised as you "haven't run into that", or "don't care about that", or "it's fine for me", which are not good enough answers for merging something into master.

It is worth noting, too, that:
- @whyrusleeping put a huge amount of effort into getting this and your other PRs to a state where we could merge them. You have dismissed that effort and attention many times, including in this PR.
- I put a lot of time into coordinating your communications with others and getting multiple people to help you improve your code, your abilities as an engineer, and helping you learn.
- A lot of of your designs (and code) will make it into master with their functionality intact.

_you said:_

> For those of you who have not figured out what is going on yet: this p.r. is unlikely to be merged as is and instead the core developers decided on a much more basic implementation. 

When you first set out to implement filestore I discouraged you from doing it. I told you that this is really hard to get right, that I had plans for implementing it in the future, and that we have very strong UX and systems constraints to satisfy, and that we need implementations to be clear and good.  I warned you that we would not merge work that did not satisfy those. (which translates to rejecting PRs that do not satisfy those strong requirements) You chose to proceed anyway.

_you said:_
> I was not included when the decision was made and am in general not at all happy with how the p.r. was handled. 

You absolutely were involved in this process, all the way through. I have spoken to you directly multiple times about this. You simply disagree with the conclusion. 

We have put a lot of effort into encouraging you, supporting you, and trying to make this PR acceptable. @whyrusleeping spent dozens of hours reviewing your code and coaching you for months in order to improve these PRs. He also spent a lot of time coaching you in basic engineering and programming issues, instead of just rejecting your code or rewriting it himself. In fact, he spent more time doing that than it would have taken him to implement this whole set of features on his own, and he did so out of respect for you and your work, and out of a desire to help you improve. You met his attempts to improve the code with resistance and hostility, and dismissed his efforts to help you. You often did this through time-consuming private IRC exchanges or calls. I would have preferred for those conversations to be in the public IRC channels or on the PR, but we respected your communication medium preference.

_you said:_
> With the way things are going it does not look like most of the functional of this branch will ever make it into IPFS.

This is completely false. Much of the designs or code that you made _will_ make it into the code base soon, and the functionality _will_ be supported. The basic filestore implementation that _will be_ merged is entirely based on the work you did for these PRs. The `ipfs-pack` tool is entirely based on the work you did for these PRs. Both add restrictions to improve the UX and avoid some of the systems issues. And, after @whyrusleeping re-implemented the lowest layer based on my and your designs last month, I invited you to move your existing code + functionality on top of the new base implementation. You added some of those changes to [this new filestore pr](https://github.com/ipfs/go-ipfs/pull/3629). More code you adapted from the original PR is in [this new PR you created](https://github.com/ipfs/go-ipfs/pull/3653), which will be merged soon.



Separately, I ask that -- if you wish to continue working with me and others on IPFS -- you pay very close attention to:

- the time people spend helping you
- the quality and correctness implementations must meet
- the correctness, systems, and UX concerns people bring up
- your wording and claims, which in this issue are unfair and misleading
- respecting the thoughts and decisions of others
- respecting the systems, UX, and code quality concerns of the project


In any case, thanks very much for all the hard work you, @whyrusleeping, and many others have put into this feature set. I think it is shaping up to be something that will add a lot of value to users and avoid a lot of the problems we've been concerned by, even if it has taken longer to land that many would have liked.

kevina commented:
@jbenet I apologize for my choice of wording.  At the moment do not know how to properly respond to your comments.  Your comments are not something that I think should of been publicly posted, I apologize if I am in the wrong here.  I would fell a lot better if my last comment and all the following comments includes yours is deleted and we deal with this matter privately.  For now I am locking the conversation as that can easily be undone.  I apologize if I am in he wrong and if I am fell free to unlock it.

I can be reached privately at k@kevina.org.

kevina commented:
@jbenet I apologize for my choice of wording and can see how they can be misleading. At the moment I do not know how to properly respond to your comments.  Some of your comments, which are of a personal nature, are not something that I think should of been publicly posted, I apologize for if I am in the wrong here.  I would fell a lot better if my last comment and all the following comments including yours is deleted and we deal with this matter privately.  For now I am locking the conversation as that can easily be undone.  I apologize if I am in the wrong here, and if I am fell free to unlock it.

I can be reached privately at k@kevina.org.

whyrusleeping commented:
:raised_hands: :fireworks: #3629 has been merged, basic filestore functionality is in master. Go try it out :)
